### PR TITLE
[Provenance QA] Carbon provenance implementation

### LINKED ISSUE #1936: [retire] Carbon provenance implementation
### Related docs & links

[Shipped Figma Design](https://www.figma.com/file/7N5csmr1NPeKefGh5MuGeh/Transfer-History?type=design&node-id=1-90&mode=design&t=rDcTnXPNjtix3jl0-0)

Related subgraph work: https://github.com/KlimaDAO/klima-subgraph/pull/75

[Design/ discovery discord thread ](https://discord.com/channels/1135827610434289727/1173629852050526289)

### Requirements & user journey

- In the figma design file, find the DEVS START HERE section
- Note the use of the material-ui react timeline component (https://mui.com/material-ui/react-timeline/)
- User journey starts from either a retirement receipt page or retirement receipt pdf
- Also not the any design differences for desktop vs mobile (included in file)

**Retirement Receipt Page**

- Remove the COPY LINK button in the current implementation
Screenshot 2023-11-29 at 2 37 31 PM

- As per the new [design](https://www.figma.com/file/7N5csmr1NPeKefGh5MuGeh/Transfer-History?type=design&node-id=18-1037&mode=design&t=rDcTnXPNjtix3jl0-4), move `VIEW ON POLYGONSCAN` button to the left and add a `CARBON PROVENANCE` button
Screenshot 2023-11-29 at 2 40 38 PM

**PDF Receipt Page**
- As per the new [design](https://www.figma.com/file/7N5csmr1NPeKefGh5MuGeh/Transfer-History?type=design&node-id=18-1131&mode=design&t=rDcTnXPNjtix3jl0-4), add a View carbon provenance link to the exported pdf
Screenshot 2023-11-29 at 2 43 30 PM

**Carbon Provenance Page (NEW)**
- This is a new page
- Will need to settle on a URL structure. Suggest https://www.carbonmark.com/retirements/0x087a7afb6975a2837453be685eb6272576c0bc06/6/provenance (so adding /provenance) but please confirm with @Atmosfearful for his recommendation
- Implement the page components as per the [design](https://www.figma.com/file/7N5csmr1NPeKefGh5MuGeh/Transfer-History?type=design&node-id=1-90&mode=design&t=rDcTnXPNjtix3jl0-4)
- _Note that the green checkmark beside an 0x address indicates a verified wallet which is future functionality and not supported yet for this MVP_

_Notes_:
- See some details on subgraph endpoint here https://github.com/KlimaDAO/klima-subgraph/pull/75
- You will likely need to sync and have a discussion with @cujowolf to fully understand the details of the endpoint and the data that is provided, which will look something like:
image

- Sample query
`[https://api.thegraph.com/subgraphs/name/klimadao/polygon-digital-carbon/graphql?query=query+MyQuery+%7B%0A++provenanceRecords(%0A++++where%3A+%7BtransactionHash%3A+"0x88bb718a2556970081e9cd6a8f17af2dffaabe1a05f91dc47cf477c3425225b7"%7D%0A++)+%7B%0A++++id%0A++++originalAmount%0A++++sender%0A++++receiver%0A++++token%0A++++transactionType%0A++++priorRecords(orderBy%3A+createdAt%2C+orderDirection%3A+desc)+%7B%0A++++++sender%0A++++++receiver%0A++++++transactionType%0A++++++originalAmount%0A++++++remainingAmount%0A++++%7D%0A++%7D%0A%7D](https://api.thegraph.com/subgraphs/name/klimadao/polygon-digital-carbon/graphql?query=query+MyQuery+%7B%0A++provenanceRecords(%0A++++where%3A+%7BtransactionHash%3A+%220x88bb718a2556970081e9cd6a8f17af2dffaabe1a05f91dc47cf477c3425225b7%22%7D%0A++)+%7B%0A++++id%0A++++originalAmount%0A++++sender%0A++++receiver%0A++++token%0A++++transactionType%0A++++priorRecords(orderBy%3A+createdAt%2C+orderDirection%3A+desc)+%7B%0A++++++sender%0A++++++receiver%0A++++++transactionType%0A++++++originalAmount%0A++++++remainingAmount%0A++++%7D%0A++%7D%0A%7D)`

### PR DESCRIPTION
**Pull request by biwano**
## Description

~~This should not be merged before #2034~~ :merged

- API: Adds 2 retirements routes
- UI: Adds a View carbon provenance button on the single retirements page
- UI: Adds a View carbon provenance link on the generated certificate
- UI: Adds a Carbon provenance page

Todo:

 - [x] Write the algorithm to find the different routes from the carbon provenance data => Feels convoluted. I am not sure if we should try to do that.
 - [x] Fix the issue with fonts being generaly too big
 - [x] Check with @cujowolf  if we can add the transaction hash into `polygon-digital-carbon/Retire` model (pending ICR release)
 - [x] Check with @cujowolf if the value of `polygon-digital-carbon/CarbonCredit/id` is correct (does not look like what Codey is expecting)
 - [x] Check with @cujowolf  the `polygon-digital-carbon/KlimaRetire/id` 0x087a7afb6975a2837453be685eb6272576c0bc0000000006 instead of 0x087a7afb6975a2837453be685eb6272576c0bc0600000000 ? => This is how the graph concat bytes
 - [x] Add a transition to the collapsible effect
 - [x] Nock Tests for the `retirements/klima/:account_id/:retirement_index` route
 - [x] Nock Tests for the `retirements/klima/:account_id/:retirement_index/provenance` route
 - [x] Postman tests
 - [x] Link to Verra retirement page

## Related Ticket

Closes #1936 

## Notes For QA

Specific pages, components or journeys that might be affected:
- https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0x087a7afb6975a2837453be685eb6272576c0bc06/6/provenance
- https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0x087a7afb6975a2837453be685eb6272576c0bc06/6 => click on Download PDF and click on Carbon Provenance

Relevant preview URLs:
- 

Other notes:
- The "View carbon provenance link" in the certificate does not work because it points at the production environment which does not have this feature yet.

### CODE REVIEW

**Review by jabby09 — CHANGES_REQUESTED** [review]
@biwano as noted in discord, some retirement pages are returning a 404 page

For example: https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0xea4d28f8218f44c84c8d7cbae5deac41acf66ced/7/provenance

or https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0x087a7afb6975a2837453be685eb6272576c0bc06/2/provenance (MOSS retirement)

**Comment by 0xtapi** [comment]
> @biwano as noted in discord, some retirement pages are returning a 404 page
> 
> For example: https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0xea4d28f8218f44c84c8d7cbae5deac41acf66ced/7/provenance
> 
> or https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0x087a7afb6975a2837453be685eb6272576c0bc06/2/provenance (MOSS retirement)

I found one that returns a 500 (server error), not a 404: https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0xca2b532dd6fbff14430b4ff2cfb2823eeb1245c8/1/provenance

Not much in the console:
Screenshot 2023-12-18 122000

All other provenance links in my retirements load and refresh without issue, but this one keeps returning a 500.

**Comment by biwano** [comment]
> > @biwano as noted in discord, some retirement pages are returning a 404 page
> > For example: https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0xea4d28f8218f44c84c8d7cbae5deac41acf66ced/7/provenance
> > or https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0x087a7afb6975a2837453be685eb6272576c0bc06/2/provenance (MOSS retirement)
> 
> I found one that returns a 500 (server error), not a 404: https://carbonmark-git-1936carbonprovenanceimplementation-klimadao.vercel.app/retirements/0xca2b532dd6fbff14430b4ff2cfb2823eeb1245c8/1/provenance
> 
> Not much in the console: Screenshot 2023-12-18 122000
> 
> All other provenance links in my retirements load and refresh without issue, but this one keeps returning a 500.

This one is fixed by [18354a1](https://github.com/KlimaDAO/klimadao/pull/2006/commits/18354a1f6df0e0e0a18f589ca4ac37b92ed73549)

There was no serial number for the bridging transaction of this one. I do not know if this is normal. 
@cujowolf Sorry to bother you all the time but I do not know the business rules around this stuff.

**Comment by 0xtapi** [comment]
I did some more testing with my retirements (eight in total) and checked them for provenance links on the page and provenance links in the PDFs. Two things I noticed (both minor or not directly related to this):

1. Incorrect values in the PDF, see:
Screenshot 2024-01-15 122548

I assume this is related to https://github.com/KlimaDAO/klimadao/issues/2123 ?

2. The URLs in the PDFs point to the production environment, where provenance isn't implemented yet, so all URLs return 404s. If I replace the URL with the preview link URL, it's working as expected. So this will probably resolve itself once it's in production.

**Comment by biwano** [comment]
> I did some more testing with my retirements (eight in total) and checked them for provenance links on the page and provenance links in the PDFs. Two things I noticed (both minor or not directly related to this):
> 
> 1. Incorrect values in the PDF, see:
>    Screenshot 2024-01-15 122548
> 
> I assume this is related to #2123 ?
Yes
> 
> 2. The URLs in the PDFs point to the production environment, where provenance isn't implemented yet, so all URLs return 404s. If I replace the URL with the preview link URL, it's working as expected. So this will probably resolve itself once it's in production.
Yes

**Review by jabby09 — CHANGES_REQUESTED** [review]
@biwano mobile version of the page needs an overhaul, lot of UI alignment, padding issues

IMG_542B6DD0794C-1
IMG_CEB68D02F8DE-1

**Comment by biwano** [comment]
> @biwano mobile version of the page needs an overhaul, lot of UI alignment, padding issues
> 
> IMG_542B6DD0794C-1 IMG_CEB68D02F8DE-1

Sorry I completely ignored the mobile designs. Please give it another go.

**Review by jabby09 — CHANGES_REQUESTED** [review]
@biwano getting closer
just this top white component that does not match [mobile design](https://www.figma.com/file/7N5csmr1NPeKefGh5MuGeh/Carbon-Provenance?type=design&node-id=1-1336&mode=design&t=UdsVpQMleBmI4VbN-4)

IMG_01FCF7C495E3-1

**Comment by biwano** [comment]
> @biwano getting closer just this top white component that does not match [mobile design](https://www.figma.com/file/7N5csmr1NPeKefGh5MuGeh/Carbon-Provenance?type=design&node-id=1-1336&mode=design&t=UdsVpQMleBmI4VbN-4)
> 
> IMG_01FCF7C495E3-1

I fixed the User information not displaying corectly the name and avatar.
If it was about the topbar though @theprofessora told me not to implement it because it will be part of the UI redesign work.

Let me know if it waqs something else.

**Review by jabby09 — CHANGES_REQUESTED** [review]
@biwano thanks for addressing the avatar / account. We have related issues on the receipt page itself captured in #2133 and #2134 

The mobile issue I was referring to is the width of the component.

So design looks like this:
Screenshot 2024-01-27 at 10 41 42 AM

And current implementation like this:
IMG_EAD9F7EC7EF6-1

**Comment by biwano** [comment]
> @biwano thanks for addressing the avatar / account. We have related issues on the receipt page itself captured in #2133 and #2134
> 
> The mobile issue I was referring to is the width of the component.
> 
> So design looks like this: Screenshot 2024-01-27 at 10 41 42 AM
> 
> And current implementation like this: IMG_EAD9F7EC7EF6-1

I am blind! You can give it another go now.

**Review by jabby09 — APPROVED** [review]

File: carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  ) : (
+    <></>
+  );
+
+  return (
+    <div className={styles.wrapper}>
+      <div className={styles.header}>
+        <RetirementCardItem
+          title={t`Amount`}
+          icon={<ParkOutlined fontSize="large" />}
```
Comment (sprrwhwk) [inline carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx:84]: icon color should be `#626266`

**Review by sprrwhwk — CHANGES_REQUESTED** [review]
Great work, everything seems to work nicely. I found some visual problems that should be straightforward fixes.

Additionally, the icons here shouldn't have a shadow and the `5,000 tonnes` background should be white:

Screenshot 2024-01-31 at 11 56 23

File: carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+  BalanceOutlined,
+  ChangeCircleOutlined,
+  DeviceHub,
+  East,
+  KeyboardArrowDown,
+  KeyboardArrowUp,
+  Park,
+  Token,
+} from "@mui/icons-material";
+import InfoOutlinedIcon from "@mui/icons-material/InfoOutlined";
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx]: Done in 299db84b9

File: carbonmark/components/Quantity/index.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+type Props = {
+  quantity: number;
+  locale?: string;
+  className?: string;
+};
+
+/**
+ * Display a quantity of credits in a tagged form
+ * FIXME: There is factorization potential with tag components (Quantity/Vintage/Category)
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/Quantity/index.tsx:16]: If you look at the Vintage and Category components, the code is very similar.

File: carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                          <Text t="responsiveBody2">
+                            <A
+                              href={`${verra.appSearch}?programType=ISSUANCE&exactResId=${projectId}`}
+                            >
+                              <Trans>View issuance on Verra</Trans>
+                            </A>
+                          </Text>
+                          <TextInfoTooltip
+                            align="start"
+                            tooltip={t`On the Verra page, search (CTRL+F) for the specific serial number.`}
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx:281]: @theprofessora

File: carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                      <div className={styles.contentFooter}>
+                        <Quantity quantity={record.originalAmount} />
+                        <Text t="responsiveBody1">
+                          {concatAddress(record.sender)}
+                        </Text>
+                        <Text t="responsiveBody1" color="lightest">
+                          via Carbonmark
+                        </Text>
+                      </div>
+                      <Divider
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx:225]: Done in 56b2168ef

File: carbonmark/components/SocialLinks/index.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * @returns
+ */
+export const SocialLinks: FC<Props> = (props) => {
+  return (
+    <div className={styles.socialLinks}>
+      <div className="buttons">
+        <TweetButton title={props.twitterTitle} tags={props.twitterTags} />
+        <FacebookButton />
+        <LinkedInButton />
+        <CopyValueButton />
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/SocialLinks/index.tsx]: Done in b0f3a1993

File: carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+          text={t`${formattedAmount} Tonnes`}
+        />
+        <RetirementCardItem title={t`By`} icon={ByIcon} text={by} />
+        <RetirementCardItem
+          title={t`Date`}
+          icon={<Today fontSize="large" />}
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx:88]: Done in 237848e85

File: carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx (hunk 1 above)
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx:82]: Done in 237848e85

**Comment by biwano** [comment]
> Great work, everything seems to work nicely. I found some visual problems that should be straightforward fixes.
> 
> Additionally, the icons here shouldn't have a shadow and the `5,000 tonnes` background should be white:
> 
> Screenshot 2024-01-31 at 11 56 23

Done in 56a4e7430

**Review by sprrwhwk — APPROVED** [review]
Looks great now. Good job! 💪

**Review by 0xemc — CHANGES_REQUESTED** [review]
**First pass (UI only)**

This looks awesome, very nice PR 👍.

A couple of small things I noticed:

image

1. `formattedAmount` Not sure if this is an artifact of using the staging environment?
2. We need some wrapping of text in the mobile versions (see attached screenshot)

File: carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+const getFormattedDate = (timestamp: number, locale = "en") => {
+  const dateObj = new Date(timestamp * 1000); //expects milliseconds
+  return new Intl.DateTimeFormat(locale, {
+    day: "2-digit",
+    month: "2-digit",
+    year: "numeric",
+  }).format(dateObj);
+};
+
+/** Helper that formats retirement information */
```
Comment (0xemc) [inline carbonmark/components/pages/Retirements/RetirementProvenance/RetirementCard/index.tsx:29]:
```suggestion
```

Incorrect jsdoc

File: carbonmark/components/pages/Retirements/RetirementProvenance/index.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+ */
+type ProvenanceList = Record<string, ProvenanceRecord[]>;
+
+/**
+ * Splits Provenance records in different streams
+ * Each stream tracks the provenance of the credits received by the last sender
+ * @param provenance
+ * @returns
+ */
+const splitProvenance = (records: ProvenanceRecord[]) => {
```
Comment (0xemc) [inline carbonmark/components/pages/Retirements/RetirementProvenance/index.tsx:28]: Again, would be great in a `.utils` file

File: carbonmark/components/pages/Retirements/RetirementProvenance/index.tsx [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+  if (records.length <= 3) return [records];
+  // Sender of the last trasfer
+  const lastSender = records[1].sender;
```
Comment (0xemc) [inline carbonmark/components/pages/Retirements/RetirementProvenance/index.tsx:31]: Are we sure the records will always be returned in this order?

File: carbonmark/pages/retirements/[beneficiary]/[retirement_index]/provenance/index.tsx [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  provenance: GetRetirementsIdProvenanceQueryResponse;
+  nameserviceDomain: string | null;
+  retirementIndex: string;
+  beneficiaryAddress: string;
+}
+
+// second param should always be a number
+const isNumeric = (value: string) => {
+  return /^\d+$/.test(value);
+};
```
Comment (0xemc) [inline carbonmark/pages/retirements/[beneficiary]/[retirement_index]/provenance/index.tsx:38]: please place in `string.utils.ts`

Comment (biwano) [reply carbonmark/pages/retirements/[beneficiary]/[retirement_index]/provenance/index.tsx]: Done in a08865c8a

File: carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+    locale: locale || "en",
+  });
+
+  /** Returns the relevant record info for a given record */
+  const recordInfo = (record: ProvenanceRecord) => {
+    if (Object.keys(RECORDS_INFO).includes(record.transactionType))
+      return RECORDS_INFO[record.transactionType as keyof typeof RECORDS_INFO];
+  };
+
+  /** Return the correct stylefor a given record */
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx]: Done in d4d66f710

File: carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+ * @param props
+ * @returns
+ */
+export const ProvenanceComponent = (props: ProvenanceComponentProps) => {
+  const { locale } = useRouter();
+  const [showTransfers, setShowTransfers] = useState<boolean>(false);
+
+  // Do not display anything if we do not have at least a record
+  const lastRecord = props.records[0];
+  if (!lastRecord) return <></>;
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx]: Done in d4d66f710

File: carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+  }).format(dateObj);
+  const date = new Intl.DateTimeFormat(locale, {
+    timeZone: "UTC",
+    day: "numeric",
+    month: "long",
+    year: "numeric",
+  }).format(dateObj);
+
+  return `UTC ${time} | ${date}`;
+};
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx]: Done in d4d66f710

File: carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+  /** Styles record items to trigger the transition effect */
+  const recordStyle = (record: ProvenanceRecord) => {
+    return `${styles.timelineItem} ${
+      recordVisible(record)
+        ? styles.timelineItemVisible
+        : styles.timelineItemHidden
+    }`;
+  };
+  const projectId = props.retirement.credit?.projectId.split("-")[1];
```
[the remark this answers is not in the document]
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/ProvenanceComponent/index.tsx]: CreditId is defined in the `carbonmark-api` module not the `lib` module I propose we do that refactoring in another PR.

File: carbonmark/components/pages/Retirements/RetirementProvenance/index.tsx (hunk 9 above)
Comment (biwano) [reply carbonmark/components/pages/Retirements/RetirementProvenance/index.tsx:28]: Done in 87b97bd1f

**Review by 0xemc — APPROVED** [review]

**Pull request merged**